jsonschema-core.md

```diff
@@ -2071,71 +2071,44 @@ Omitting this keyword has the same assertion behavior as an empty schema.
 
-This specification defines three output formats. See the "Output Structure"
-section for the requirements of each format.
-
-- *Flag*: A boolean which simply indicates the overall validation result with no
-  further details.
-- *List*: Provides validation information in a flat list structure.
-
-Hierarchical: Provides validation information in a hierarchical structure that
-follows the evaluation paths generated while processing the schema.
-
-An implementation MUST provide the "flag" format and SHOULD provide at least one
-of the "list" or "hierarchical" formats. Implementations SHOULD specify in their
-documentation which formats they support.
-
-### Minimum Information
-
-Beyond the simplistic "flag" output, additional information is useful to aid in
-debugging a schema or instance. Each sub-result SHOULD contain the information
-contained within this section at a minimum.
-
-A single object that contains all of these components is considered an output
-unit.
-
-Implementations MAY elect to provide additional information.
-
-#### Evaluation path
-
-The evaluation path to the schema object that produced the output unit. The
-value MUST be expressed as a JSON Pointer, and it MUST include any by-reference
-applicators such as `$ref` or `$dynamicRef`.[^13]
-
-[^13]: The schema may not actually have a value at the location indicated by
-this pointer. It is provided as an indication of the traversal path only.
+The evaluation path is the set of keys, starting from the schema root, through
+which evaluation passes to reach the schema object that produced a specific
+result. The value MUST be expressed as a JSON Pointer, and it MUST include any
+by-reference applicators such as `$ref` or `$dynamicRef`.
 
 ```
 /properties/width/$ref/allOf/1
 ```
 
-Note that this pointer may not be resolvable by the normal JSON Pointer process
-due to the inclusion of these by-reference applicator keywords.
+Note that this pointer may not be resolvable on the root schema by the normal
+JSON Pointer process. It is intended as an indication of the traversal path
+only.
 
-The JSON key for this information is "evaluationPath".
+When represented in JSON, the key for this information MUST be "evaluationPath".
 
-#### Schema Location
+### Schema Location
 
-The absolute, dereferenced location of the schema object that produced the
-output unit. The value MUST be expressed using the canonical IRI of the relevant
-schema resource plus a JSON Pointer fragment that indicates the schema object
-that produced the output. It MUST NOT include by-reference applicators such as
-`$ref` or `$dynamicRef`.[^14]
+The schema location is the absolute, dereferenced location of the schema object
+that produced a result. The value MUST be expressed using the canonical IRI of
+the relevant schema resource plus a JSON Pointer fragment that indicates the
+schema object that produced the output. It MUST NOT include by-reference
+applicators such as `$ref` or `$dynamicRef`.[^14]
```

Comment:
Saying that it MUST NOT include `$ref` is, I think, a little more restrictive than is intended. It makes sense for the pointer to include a `$ref`, but only as the terminating value. Since 2019-09, `$ref` is a normal keyword and it make sense for schema location to point to that keyword as much as any other.

Reply:
There's not a schema at a `$ref`, only a string (URI).

```json
{
  "type": "array",
  "items": { "$ref": "foo.json" }
}
```

The schema location indicates the schema object.  There are only two schema objects here:

- the root
- the `items` subschema

If an evaluation result comes from _inside_ the ref, then the schema location is `foo.json` or some subschema contained by it.

Thus, "schema location" never needs to carry a reference keyword.

Reply:
I was thinking that "schema location" meant the location _in_ the schema rather than the location _of_ a schema, but reading more carefully, I see I had that wrong. However, if this section is intended to be a set of universal concepts output format authors can use when creating custom output formats, it seems a bit limiting to not provide a concept that allows them to point to a keyword. I won't hold up this PR for that. It can be worked out later if necessary.

Reply:
Yeah, the concepts currently listed are the ones that are currently used by the (now split) output spec.  I don't mind additional things being added, but I think I'd want to see them being used first.  You know, see the term in another output spec, and then import it into this spec to standardize it so that other output specs can use the same term.